src/dataEntities/appointment.ts

```diff
@@ -1,149 +1,328 @@
+        if (!this.mEventFilter) {
+            this.mEventFilter = this.getEventFilter();
+        }
+        return this.mEventFilter;
+    }
+    private mEventFilter: ethers.EventFilter;
+    private getEventFilter(): ethers.EventFilter {
+        // the abi is in human readable format, we can parse it with ethersjs
+        // then check that it's of the right form before separating the name and inputs
+        // to form topics
+
+        const eventInterface = new ethers.utils.Interface([this.eventABI]);
+        if (eventInterface.abi.length !== 1) throw new PublicDataValidationError("Invalid ABI. ABI must specify a single event."); // prettier-ignore
+        const event = eventInterface.abi[0];
+
+        if (event.type !== "event") throw new PublicDataValidationError("Invalid ABI. ABI must specify an event.");
+
+        const name = eventInterface.abi[0].name;
+        const inputs = eventInterface.abi[0].inputs;
+
+        // we encode within the data which inputs we'll be filtering on
+        // so the first thing encoded is an array of integers representing the
+        // indexes of the arguments that will be used in the filter.
+        // non specified indexes will be null
+        const indexes: BigNumber[] = ethers.utils.defaultAbiCoder.decode(["uint256[]"], this.eventArgs)[0];
+        const namedInputs = indexes.map(i => i.toNumber()).map(i => inputs[i]);
+        const decodedInputs = ethers.utils.defaultAbiCoder
+            .decode(["uint256[]"].concat(namedInputs.map(i => i.type)), this.eventArgs)
+            .slice(1);
+
+        const topics = eventInterface.events[name].encodeTopics(decodedInputs);
         return {
-            ...this,
-            startBlock: this.startBlock,
-            endBlock: this.endBlock
+            address: this.contractAddress,
+            topics
         };
     }
+    /**
+     * Th packed representation for the
+     */
+    public solidityPacked() {
+        return ethers.utils.solidityPack(
+            [
+                "address",
+                "address",
+                "uint",
+                "uint",
+                "uint",
+                "uint",
+                "uint",
+                "bytes",
+                "uint",
+                "uint",
+                "uint",
+                "bytes",
+                "bytes",
+                "bytes",
+                "bytes32"
+            ],
+            [
+                this.contractAddress,
+                this.customerAddress,
+                this.startBlock,
+                this.endBlock,
+                this.challengePeriod,
+                this.customerChosenId,
+                this.jobId,
+                this.data,
+                this.refund,
+                this.gas,
+                this.mode,
+                ethers.utils.toUtf8Bytes(this.eventABI),
+                this.eventArgs,
+                this.postCondition,
+                this.paymentHash
+            ]
+        );
+    }
```

Comment:
If ethers.js doesn't have a way of encoding lists of arguments as arrays of
```
{
 type: "<>",
 value: <>
}
```
or something similar, perhaps we should write one!

Reply:
Added a helper function called "groupTuples"